apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClientBuilder.cs

```diff
@@ -0,0 +1,153 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.PubSub.V1;
+
+/// <summary>
+/// Builder class for <see cref="PublisherClient"/> to provide simple configuration of credentials,
+/// endpoint, client count, publication settings etc.
+/// </summary>
+public sealed class PublisherClientBuilder : ClientBuilderBase<PublisherClient>
+{
+    private static readonly GrpcChannelOptions s_unlimitedSendReceiveChannelOptions = GrpcChannelOptions.Empty
+        .WithMaxReceiveMessageSize(int.MaxValue)
+        .WithMaxSendMessageSize(int.MaxValue)
+        // Set max metadata size to 4 MB i.e., 4194304 bytes.
+        .WithCustomOption("grpc.max_metadata_size", 4194304);
+
+    /// <summary>
+    /// The name of the topic that the publisher publishes to.
+    /// This must be non-null by the time <see cref="Build"/> or <see cref="BuildAsync(CancellationToken)"/> is called.
+    /// </summary>
+    public TopicName TopicName
+    {
+        get; set;
+    }
+
+    private int? _clientCount;
+
+    /// <summary>
+    /// The number of <see cref="PublisherServiceApiClient"/>s to create and use within a <see cref="PublisherClient"/> instance.
+    /// If this is null, the number of clients created will depend on the processor count.
+    /// </summary>
+    public int? ClientCount
+    {
+        get => _clientCount;
+        set
+        {
+            if (value is int count)
+            {
+                // Fairly arbitrary upper limit.
+                _clientCount = GaxPreconditions.CheckArgumentRange(count, nameof(value), 1, 256);
+            }
+            else
+            {
+                _clientCount = value;
+            }
+        }
+    }
+
+
+    /// <summary>
+    /// Additional settings for batching, message ordering etc. Default settings will be used if this is null.
+    /// </summary>
+    public PublisherClient.Settings Settings { get; set; }
+
+    /// <summary>
+    /// The settings to use when creating<see cref= "PublisherServiceApiClient" /> instances.
+    /// Default settings will be used if this is null.
+    /// </summary>
+    public PublisherServiceApiSettings ApiSettings { get; set; }
+
+    /// <summary>
+    /// Specifies how to respond to the presence of emulator environment variables.
+    /// </summary>
+    /// <remarks>
+    /// This property defaults to <see cref="EmulatorDetection.None"/>, meaning that
+    /// environment variables are ignored.
+    /// </remarks>
+    public new EmulatorDetection EmulatorDetection
+    {
+        get => base.EmulatorDetection;
+        set => base.EmulatorDetection = value;
+    }
+
+    /// <summary>
+    /// Creates a new instance with default settings. <see cref="TopicName"/> must be set
+    /// before calling either <see cref="Build"/> or <see cref="BuildAsync"/>.
+    /// </summary>
+    public PublisherClientBuilder() : base(PublisherServiceApiClient.ServiceMetadata)
+    {
+    }
+
+    /// <summary>
+    /// 
```

Comment:
Do we need docs here and below?

Reply:
Yes indeed - thanks for catching this. I'll fix these and then merge.